packages/ember/tests/routing/substates_test.js

```diff
@@ -1,1042 +1,880 @@
+      assert.throws(() => {
+        this.visit('/foo');
+      }, (err) => err.msg === 'did it broke?', 'it broke');
+      let text = this.$('#app').text();
 
-  Router.map(function() {
-    this.route('grandma', function() {
-      this.route('mom', { resetNamespace: true }, function() {
-        this.route('sally');
-      });
+      assert.equal(text, 'FOO ERROR: did it broke?',
+        'foo.index_error was entered'
+      );
     });
-  });
-
-  App.ApplicationController = Controller.extend();
-
-  App.MomSallyRoute = Route.extend({
-    model() {
-      step(1, 'MomSallyRoute#model');
-
-      return RSVP.reject({
-        msg: 'did it broke?'
-      });
-    },
-
-    actions: {
-      error(err) {
-        step(2, 'MomSallyRoute#error');
-        return true;
-      }
-    }
-  });
-
-  throws(
-    function() {
-      bootApplication('/grandma/mom/sally');
-    },
-    function(err) {
-      return err.msg === 'did it broke?';
-    },
-    'Correct error was thrown'
-  );
+  }
 });
 
-QUnit.test('Handled errors that are thrown through rejection aren\'t swallowed', function() {
-  expect(4);
-
-  let handledError;
-
-  templates['grandma'] = 'GRANDMA {{outlet}}';
+moduleFor('Loading/Error Substates - globals mode app', class extends AutobootApplicationTestCase {
+  /*
+   * When an exception is thrown during the initial rendering phase, the
+   * `visit` promise is not resolved or rejected. This means the `applicationInstance`
+   * is never torn down and tests running after this one will fail.
+   *
+   * It is ugly, but since this test intentionally causes an initial render
+   * error, it requires globals mode to access the `applicationInstance`
+   * for teardown after test completion.
+   *
+   * Application "globals mode" is trigged by `autoboot: true`. It doesn't
+   * have anything to do with the resolver.
+   *
+   * We should be able to fix this by having the application eagerly stash a
+   * copy of each application instance it creates. When the application is
+   * destroyed, it can also destroy the instances (this is how the globals
+   * mode avoid the problem).
+   *
+   * See: https://github.com/emberjs/ember.js/issues/15327
+   */
+  ['@test Rejected promises returned from ApplicationRoute transition into top-level application_error'](assert) {
+    assert.throws(() => {
+      this.runTask(() => {
+        this.createApplication();
+        this.add('route:application', Route.extend({
+          init() {
+            this._super(...arguments);
+          },
+          model() {
+            return RSVP.reject({ msg: 'BAD NEWS BEARS' });
+          }
+        }));
+
+        this.addTemplate('application_error', `
+          <p id="toplevel-error">TOPLEVEL ERROR: {{model.msg}}</p>
+        `);
+      })
+    }, (err) => err.msg === 'BAD NEWS BEARS', 'it went poorly');
+
+    let text = this.$('#toplevel-error').text();
+    assert.equal(text, 'TOPLEVEL ERROR: BAD NEWS BEARS',
+      'toplevel error rendered'
+    );
```

Comment:
I can add that back in, my $0.02 is that you're testing behavior well tested in the basic routing test with that second assertion at the cost of pretty gnarly testing setup.